pykafka/producer.py

```diff
@@ -237,7 +428,167 @@ def produce(self, messages):
+
+    :ivar lock: The lock used to control access to shared resources for this
+        queue
+    :ivar flush_ready: A condition variable that indicates that the queue is
+        ready to be flushed via requests to the broker
+    :ivar slot_available: A condition variable that indicates that there is
+        at least one position free in the queue for a new message
+    :ivar queue: The message queue for this broker. Contains messages that have
+        been supplied as arguments to `produce()` waiting to be sent to the
+        broker.
+    :type queue: collections.deque
+    :ivar messages_inflight: A counter indicating how many messages have been
+        enqueued for this broker and not yet sent in a request.
+    :type messages_inflight: int
+    :ivar producer: The producer to which this OwnedBroker instance belongs
+    :type producer: :class:`pykafka.producer.AsyncProducer`
+    """
+    def __init__(self, producer, broker):
+        self.producer = producer
+        self.broker = broker
+        self.lock = self.producer._cluster.handler.Lock()
+        self.flush_ready = self.producer._cluster.handler.Event()
+        self.slot_available = self.producer._cluster.handler.Event()
+        self.queue = deque()
+        self.messages_inflight = 0
+
+        def queue_reader():
+            while True:
+                try:
+                    batch = self.flush(self.producer._linger_ms)
+                    if batch:
+                        self.producer._prepare_request(batch, self)
+                except Exception as e:
+                    # surface all exceptions to the main thread
+                    self.producer._worker_exception = e
+                    break
+        log.info("Starting new produce worker for broker %s", broker.id)
+        self.producer._cluster.handler.spawn(queue_reader)
+
+    def message_is_inflight(self):
+        """
+        Indicates whether there are currently any messages that have been
+            `produce()`d and not yet sent to the broker
+        """
+        return self.messages_inflight > 0
+
+    def enqueue(self, messages, acquire=True):
+        """Push messages onto the queue
+
+        :param messages: The messages to push onto the queue
+        :type messages: iterable of tuples of the form
+            `((key, value), partition_id)`
+        :param acquire: Whether to acquire and release the lock. If False,
+            this function assumes that the lock is already held by the caller
+            and raises an AssertionError if not.
+        :type acquire: bool
+        """
+        assert acquire or self.lock.locked()
+        self._wait_for_slot_available(acquire=acquire)
+        if acquire:
+            self.lock.acquire()
+        self.queue.extendleft(messages)
+        self.messages_inflight += len(messages)
+        if len(self.queue) >= self.producer._max_queued_messages:
+            if not self.flush_ready.is_set():
```

Comment:
Sounds like you're essentially suggesting that the message count thresholds for emptying the queue and declaring that it's full. Seems reasonable to me.